core/src/main/java/org/apache/druid/collections/CombiningIterable.java

```diff
@@ -51,37 +50,32 @@
     return create(
         new MergeIterable<InType>(comparator, (Iterable<Iterable<InType>>) in),
         comparator,
-        new BinaryFn<InType, InType, InType>()
-        {
-          @Override
-          public InType apply(InType arg1, InType arg2)
-          {
-            if (arg1 == null) {
-              return arg2;
-            }
-            return arg1;
+        (arg1, arg2) -> {
```

Comment:
Could you please extract `GuavaUtils.firstNonNull` and use a method reference here? The doc for that method should note that `Objects.firstNonNull()` cannot be used itself because it's one of those methods causing Guava incompatibility (see #6948). Please also add Guava's `Objects.firstNonNull()` this method to forbidden-apis.

Reply:
This pattern is in a lot of places, but this is the only place specifically where it's done to just return one or the other, not as a mechanism to bail early before doing some sort of combine. Is it still worth moving to `GuavaUtils` or should it just be a private static method of this class?

Additionally, this method seems `@Nullable` which makes it have a different contract than `Objects.firstNonNull`, which requires one of the 2 arguments not be null. Prohibiting it on this functions behalf doesn't seem necessary, so are we doing it for a reason in #6948?

Reply:
I went ahead and did this ^ with the latest commits, extracting to `GuavaUtils.firstNonNull` and adding to forbidden api